File: app/scraper/runner.py
"""Background thread that runs a scrape and exposes its progress/state.

Owns exactly one worker thread at a time. The worker opens its own sqlite
connection (db.connect is not thread-safe to share) and drives JobPortalScraper,
writing each job to the database as soon as its `scraped` event arrives, so the
UI can watch jobs come in while the scrape is still running.

Every status change bumps `version` and notifies a threading.Condition, so a
reader can block on wait_for_change() instead of polling -- that is what the
/api/scraper/stream SSE route in server.py streams to the browser.
"""

import logging
import sqlite3
import threading
from collections import deque
from datetime import datetime, timezone
from pathlib import Path
from typing import Callable, Optional

from selenium.common.exceptions import (
    InvalidSessionIdException,
    NoSuchWindowException,
    WebDriverException,
)

from app import config, db
from app.scraper.browser import create_driver
from app.scraper.portal import BrowserClosed, Cancelled, JobPortalScraper

logger = logging.getLogger(__name__)

# "cancelling" counts as running: the worker is still unwinding, so starting a new
# scrape (or clearing the jobs) stays blocked until it reaches "cancelled".
_RUNNING_STATES = {"starting", "waiting_for_login", "scraping", "cancelling"}

# How many lines of the human-readable activity log are kept (oldest dropped).
EVENT_LOG_MAX = 200

# Level for the `events` log, per portal event type; anything else is "info".
_EVENT_LEVELS = {"retry": "warn", "skipped": "error"}

_BROWSER_CLOSED_MARKERS = (
    "invalid session id",
    "not connected to devtools",
    "target window already closed",
    "no such window",
    "connection refused",
    "browser window was closed",
)


class ScraperError(Exception):
    code = "scraper_error"


class AlreadyRunning(ScraperError):
    code = "already_running"


class DatabaseNotEmpty(ScraperError):
    code = "database_not_empty"


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


def _event_message(event: dict, save_error: str = "") -> str:
    """One short, readable line for the `events` log.

    Deliberately never includes the `job` dict a `scraped` event carries: the log
    is meant to be read, and the job itself is already on its way to the database.
    `save_error` turns a `scraped` line into the reason the job did NOT land in
    the database (see ScrapeRunner._run's save_job).
    """
    kind = event.get("type", "")
    where = f"p{event.get('page')} · {event.get('index')}/{event.get('total')} · {event.get('title', '')}"
    if kind == "pages_found":
        total_pages = event.get("total_pages")
        if total_pages is None:
            return "Could not read the portal's page count; scraping anyway."
        return f"Found {total_pages} page(s), {event.get('cards_per_page') or 0} job(s) per page."
    if kind == "page":
        last_page = event.get("last_page")
        return f"Scraping page {event.get('page')}" + (f" of {last_page}." if last_page else ".")
    if kind == "retry_pass":
        return f"Retrying {event.get('cards')} skipped card(s) across {event.get('pages')} page(s)."
    if kind == "scraped":
        if save_error:
            return f"{where}: could not be saved ({save_error[:120]})."
        return f"{where}: saved."
    if kind == "retry":
        return f"{where}: retry {event.get('attempt')} ({event.get('reason', '')[:120]})."
    if kind == "skipped":
        return f"{where}: skipped ({event.get('reason', '')[:120]})."
    return kind or "event"


class ScrapeRunner:
    def __init__(
        self,
        db_path: Path | str,
        search_url: str,
        driver_factory: Callable = create_driver,
        debug_dir: Optional[Path] = None,
    ):
        self.db_path = db_path
        self.search_url = search_url
        self.driver_factory = driver_factory
        self.debug_dir = debug_dir
        # A Condition, not a plain Lock, so readers can block on wait_for_change();
        # it still guards self._status exactly like the Lock it replaced.
        self._lock = threading.Condition()
        self._version = 0
        self._cancel_event = threading.Event()
        self._thread: Optional[threading.Thread] = None
        self._driver = None
        self._status = self._initial_status()

    @staticmethod
    def _initial_status() -> dict:
        return {
            "state": "idle",
            "message": "",
            "pages_completed": 0,
            "jobs_saved": 0,
            "cards_seen": 0,
            "duplicates": 0,
            "failed": 0,
            "failed_jobs": [],
            # Final failures only, set once the run (and its retry pass) is done --
            # see JobPortalScraper._retry_failed_cards and the module docstring there.
            "anomalies": 0,
            "retries": 0,
            # Live progress: the card being scraped right now, the portal's page
            # count read up front, and total_pages x cards-on-page-1. The last two
            # stay None when the portal's pager could not be read.
            "current": None,
            "total_pages": None,
            "estimated_jobs": None,
            "events": deque(maxlen=EVENT_LOG_MAX),
            "error": None,
            "started_at": None,
            "finished_at": None,
        }

    @property
    def is_running(self) -> bool:
        with self._lock:
            return self._status["state"] in _RUNNING_STATES

    def status(self) -> dict:
        with self._lock:
            return self._snapshot()

    def wait_for_change(self, last_version: int, timeout: float = 15.0) -> tuple[int, dict]:
        """Block until the status changes, then return (version, snapshot).

        Returns at once when `last_version` is already out of date, and returns the
        unchanged version and snapshot when `timeout` passes with no change (the
        SSE route turns that into a heartbeat).
        """
        with self._lock:
            if self._version == last_version:
                self._lock.wait(timeout)
            return self._version, self._snapshot()

    def _snapshot(self) -> dict:
        """JSON-serialisable copy of the status. The lock must be held."""
        snapshot = dict(self._status)
        snapshot["events"] = list(snapshot["events"])
        snapshot["version"] = self._version
        return snapshot

    def _bump(self) -> None:
        """Mark the status as changed and wake wait_for_change. Lock must be held."""
        self._version += 1
        self._lock.notify_all()

    def _append_event(self, message: str, level: str = "info") -> None:
        """Add one line to the activity log. Lock must be held."""
        self._status["events"].append({"time": _now(), "level": level, "message": message})

    def _update(self, **fields) -> None:
        with self._lock:
            self._status.update(fields)
            self._bump()

    def _set_state(self, state: str, message: str, level: str = "info", **fields) -> None:
        """Like _update, but for a state change: the message is logged to `events`
        too, so the activity log tells the whole story of the run."""
        with self._lock:
            if state in _RUNNING_STATES and self._status["state"] == "cancelling":
                # An abort is already under way; a running state arriving late (an
                # on_ready fired while cancel() ran) must not undo it. Terminal
                # states still apply -- that is how the run ends.
                return
            self._status.update(state=state, message=message, **fields)
            if state not in _RUNNING_STATES:
                self._status["current"] = None  # nothing is being scraped any more
            self._append_event(message, level)
            self._bump()

    def _save_last_run(self, conn: sqlite3.Connection, state: str) -> None:
        """Persist the run under `state` ("running" until it ends), so the UI still
        knows an aborted or interrupted run left an incomplete job list behind,
        even after a restart. A database problem here must never fail the run."""
        status = self.status()
        try:
            db.set_last_run(
                conn,
                {
                    "state": state,
                    "started_at": status["started_at"],
                    "finished_at": status["finished_at"],
                    "pages_completed": status["pages_completed"],
                    "total_pages": status["total_pages"],
                    "jobs_saved": status["jobs_saved"],
                },
            )
        except sqlite3.Error as exc:
            logger.warning("Could not record the last run: %s", exc)

    def start(self) -> dict:
        with self._lock:
            if self._status["state"] in _RUNNING_STATES:
                raise AlreadyRunning("A scrape is already running.")
            conn = db.connect(self.db_path)
            try:
                db.init_db(conn)
                if db.count_jobs(conn) > 0:
                    raise DatabaseNotEmpty("The database already has jobs in it.")

                self._cancel_event = threading.Event()
                self._status = self._initial_status()
                self._status.update(
                    state="starting", message="Starting the scraper browser.", started_at=_now()
                )
                self._append_event(self._status["message"])
                self._bump()
                # Recorded before the worker starts: if the app is killed mid-scrape,
                # this unfinished record is all that says the job list is incomplete.
                self._save_last_run(conn, "running")
            finally:
                conn.close()

        self._thread = threading.Thread(target=self._run, name="scraper-worker", daemon=True)
        self._thread.start()
        return self.status()

    def cancel(self) -> dict:
        self._cancel_event.set()
        with self._lock:
            # Say so before the driver is quit, so the snapshot this returns (and the
            # SSE stream) already reads "cancelling" instead of "scraping"; the worker
            # only reaches "cancelled" once it has unwound, seconds later.
            if self._status["state"] in _RUNNING_STATES:
                self._set_state("cancelling", "Aborting the scrape.", level="warn")
        driver = self._driver
        if driver is not None:
            try:
                driver.quit()
            except WebDriverException as exc:
                logger.debug("Ignoring error while quitting the browser during cancel: %s", exc)
        return self.status()

    def _run(self) -> None:
        driver = None
        debug_dir = self.debug_dir if self.debug_dir is not None else config.DEBUG_DIR
        conn = db.connect(self.db_path)
        try:
            driver = self.driver_factory()
            self._driver = driver
            self._set_state(
                "waiting_for_login", "Log in to the portal in the browser window that opened."
            )

            def on_ready() -> None:
                self._set_state("scraping", "Portal login detected. Scraping jobs now.")

            def sync_accounting() -> None:
                """Copy the scraper's own counters into the status. Lock must be held.

                Live progress only -- these are pre-retry, and the retry pass only runs
                once the whole portal has been paged through (see JobPortalScraper.run),
                so "anomalies" is left alone here and only set once, from the final
                post-retry count below.
                """
                self._status["cards_seen"] = scraper.cards_seen
                self._status["duplicates"] = scraper.duplicates
                self._status["failed"] = len(scraper.failed)
                self._status["failed_jobs"] = scraper.failed[:50]

            def save_job(job: dict) -> str:
                """Write one freshly scraped job on the worker's own connection.

                Returns "" when the row landed, otherwise why it did not -- the
                scrape carries on either way, but the activity log must say so
                instead of claiming the job was saved.
                """
                try:
                    if db.insert_jobs(conn, [job]) == 1:
                        return ""
                    # The portal's own duplicate check already ran, so this is the
                    # unique index rejecting a job number we thought was new.
                    return "the database already has that job number"
                except sqlite3.Error as exc:
                    logger.warning("Could not save job %r: %s", job.get("job_number", ""), exc)
                    return str(exc)

            def on_event(event: dict) -> None:
                """Live progress from the scraper: keep the counters and the activity
                log, and save each accepted job right away so the grid can show jobs
                arriving during the scrape."""
                kind = event.get("type", "")
                save_error = save_job(event["job"]) if kind == "scraped" and event.get("job") else ""
                with self._lock:
                    if kind == "pages_found":
                        total_pages = event.get("total_pages")
                        cards_per_page = event.get("cards_per_page") or 0
                        self._status["total_pages"] = total_pages
                        self._status["estimated_jobs"] = (
                            total_pages * cards_per_page if total_pages and cards_per_page else None
                        )
                    elif kind == "card":
                        self._status["current"] = {
                            "page": event.get("page"),
                            "index": event.get("index"),
                            "total": event.get("total"),
                            "title": event.get("title", ""),
                        }
                    elif kind == "scraped":
                        if not save_error:
                            self._status["jobs_saved"] += 1
                    elif kind == "retry":
                        self._status["retries"] += 1
                    sync_accounting()
                    if kind != "card":
                        # `card` is live status only (`current` above): logging it too
                        # would spend two of the 200 lines on every single job.
                        level = "error" if save_error else _EVENT_LEVELS.get(kind, "info")
                        self._append_event(_event_message(event, save_error), level)
                    self._bump()

            def on_page(page_number: int, jobs: list) -> None:
                # Page counters only: every job here was already written to the
                # database by its own `scraped` event, in the main sweep and in the
                # retry pass alike. Inserting again would duplicate any job with a
                # blank job number, which the unique index does not cover.
                with self._lock:
                    self._status["pages_completed"] += 1
                    sync_accounting()
                    self._status["message"] = f"Finished page {page_number} ({len(jobs)} job(s))."
                    self._bump()

            scraper = JobPortalScraper(
                driver,
                self.search_url,
                cancel_event=self._cancel_event,
                page_callback=on_page,
                debug_dir=debug_dir,
                snapshot_all=config.DEBUG_SNAPSHOTS,
                ready_callback=on_ready,
                on_event=on_event,
            )
            scraper.run()

            # scraper.run() has already completed its own end-of-run retry pass, so
            # scraper.failed/anomalies are the final, authoritative counts here --
            # not the intermediate ones on_page saw while pages were still in flight.
            failed_jobs = scraper.failed
            jobs_saved = self.status()["jobs_saved"]
            self._set_state(
                "completed",
                (
                    f"Saved {jobs_saved} of {scraper.cards_seen} jobs "
                    f"({scraper.duplicates} duplicates, {len(failed_jobs)} failed)."
                ),
                cards_seen=scraper.cards_seen,
                duplicates=scraper.duplicates,
                failed=len(failed_jobs),
                failed_jobs=failed_jobs[:50],
                anomalies=len(failed_jobs),
                finished_at=_now(),
            )
        except Cancelled:
            self._set_state("cancelled", "Scrape was cancelled.", level="warn", finished_at=_now())
        except (BrowserClosed, InvalidSessionIdException, NoSuchWindowException):
            message = "Scrape was cancelled." if self._cancel_event.is_set() else "Browser window was closed."
            self._set_state("cancelled", message, level="warn", finished_at=_now())
        except Exception as exc:  # noqa: BLE001 -- top-level worker boundary, must not crash the thread
            if self._cancel_event.is_set():  # cancel() quits the browser, which surfaces as a driver error
                self._set_state("cancelled", "Scrape was cancelled.", level="warn", finished_at=_now())
            elif self._is_browser_closed(exc):
                self._set_state("cancelled", "Browser window was closed.", level="warn", finished_at=_now())
            else:
                logger.exception("Scrape failed.")
                self._set_state("failed", str(exc), level="error", error=str(exc), finished_at=_now())
        finally:
            # One write for every way the run can end -- including a driver_factory
            # that never returned a browser: every branch above has already set its
            # terminal state, so this records the run as it really finished.
            self._save_last_run(conn, self.status()["state"])
            conn.close()
            self._driver = None
            if driver is not None:
                try:
                    driver.quit()
                except WebDriverException as exc:
                    logger.debug("Ignoring error while quitting the browser during cleanup: %s", exc)

    @staticmethod
    def _is_browser_closed(exception: Exception) -> bool:
        message = str(exception).lower()
        return any(marker in message for marker in _BROWSER_CLOSED_MARKERS)

File: extension/src/background/runner.js
// Port of app/scraper/runner.py's ScrapeRunner as a plain state machine, with
// no chrome.* dependency at all (that lives in ./index.js, the service
// worker that wires this class to chrome.windows/tabs/storage/runtime.ports).
// This split is what makes the class unit-testable with an in-memory fake
// storage and transport, the same way test_runner.py drives ScrapeRunner
// directly against a FakeScraper.
//
// Execution-model mapping from the Python original:
//   * There is no worker thread here -- the "worker" is the portal content
//     script (src/content/portal.js), running JobPortalScraper in its own
//     execution context and reporting back over a chrome.runtime.Port. This
//     class only reacts to that reporting, via the handle*() methods below,
//     which stand in for runner.py's on_ready/on_event/on_page callbacks and
//     the tail of `_run`'s try/except.
//   * threading.Condition + wait_for_change() (used by the old SSE stream) is
//     replaced by subscribe(listener): every bump() calls every subscriber
//     with the fresh snapshot, which is what index.js uses to push status to
//     bridge ports, and what runner.test.js uses in place of wait_for_change.
//   * db.set_last_run has no equivalent here: this class has no access to the
//     SQLite database (that lives in the *web app's* worker, a different
//     origin entirely). Writing `last_run` is web/src/api.js's job now (see
//     its startScraper/subscribe wiring) -- this class only tracks its own
//     in-memory + chrome.storage-persisted status.
//   * Saving a scraped job is no longer synchronous (save_job in Python ran
//     on the same thread against the same sqlite connection). Each `scraped`
//     event instead becomes a persistent outbox entry handed to `transport`;
//     the jobs_saved counter and the "saved"/"could not be saved" log line
//     are only written once ack() is called for that entry -- see the module
//     docstring in the phase-5 porting brief.

// "cancelling" counts as running: a new scrape (or a re-check that the
// outbox is empty) stays blocked until the run actually reaches "cancelled".
export const RUNNING_STATES = new Set(["starting", "waiting_for_login", "scraping", "cancelling"]);

// How many lines of the human-readable activity log are kept (oldest dropped).
export const EVENT_LOG_MAX = 200;

// Level for the `events` log, per portal event type; anything else is "info".
const EVENT_LEVELS = { retry: "warn", skipped: "error" };

// Kept verbatim from Python's _BROWSER_CLOSED_MARKERS for parity, in case an
// error message (rather than a dedicated error name) is what reports a
// closed tab -- see handleError().
const BROWSER_CLOSED_MARKERS = [
  "invalid session id",
  "not connected to devtools",
  "target window already closed",
  "no such window",
  "connection refused",
  "browser window was closed",
  "extension context invalidated",
  "could not establish connection",
  "receiving end does not exist",
];

const DEFAULT_STORAGE_KEY = "coopjobs.scraper.runner";

export class ScraperError extends Error {
  constructor(message) {
    super(message);
    this.name = "ScraperError";
    this.code = "scraper_error";
  }
}

export class AlreadyRunning extends ScraperError {
  constructor(message) {
    super(message);
    this.name = "AlreadyRunning";
    this.code = "already_running";
  }
}

export class OutboxNotEmpty extends ScraperError {
  constructor(message) {
    super(message);
    this.name = "OutboxNotEmpty";
    this.code = "outbox_not_empty";
  }
}

function now() {
  return new Date().toISOString();
}

function isBrowserClosedMessage(message) {
  const text = String(message || "").toLowerCase();
  return BROWSER_CLOSED_MARKERS.some((marker) => text.includes(marker));
}

/** One short, readable line for the `events` log -- verbatim port of
 * runner.py's _event_message, including never mentioning the job dict. */
export function eventMessage(event, saveError = "") {
  const kind = event.type || "";
  const where = `p${event.page} · ${event.index}/${event.total} · ${event.title || ""}`;
  if (kind === "pages_found") {
    const totalPages = event.total_pages;
    if (totalPages === null || totalPages === undefined) {
      return "Could not read the portal's page count; scraping anyway.";
    }
    return `Found ${totalPages} page(s), ${event.cards_per_page || 0} job(s) per page.`;
  }
  if (kind === "page") {
    const lastPage = event.last_page;
    return `Scraping page ${event.page}` + (lastPage ? ` of ${lastPage}.` : ".");
  }
  if (kind === "retry_pass") {
    return `Retrying ${event.cards} skipped card(s) across ${event.pages} page(s).`;
  }
  if (kind === "scraped") {
    if (saveError) return `${where}: could not be saved (${String(saveError).slice(0, 120)}).`;
    return `${where}: saved.`;
  }
  if (kind === "retry") {
    return `${where}: retry ${event.attempt} (${String(event.reason || "").slice(0, 120)}).`;
  }
  if (kind === "skipped") {
    return `${where}: skipped (${String(event.reason || "").slice(0, 120)}).`;
  }
  return kind || "event";
}

export class ScrapeRunner {
  /**
   * @param {object} [options]
   * @param {{get(keys):Promise<object>, set(items):Promise<void>}} [options.storage]
   *   chrome.storage.local-shaped; omit for a runner that persists nothing
   *   (used by most unit tests).
   * @param {{sendJob(entry):void}} [options.transport] Handed each outbox
   *   entry as it is enqueued -- index.js's implementation forwards it down
   *   the bridge port; tests use a fake that just records calls.
   * @param {string} [options.storageKey]
   */
  constructor({ storage = null, transport = null, storageKey = DEFAULT_STORAGE_KEY } = {}) {
    this.storage = storage;
    this.transport = transport;
    this._storageKey = storageKey;

    this._version = 0;
    this._status = ScrapeRunner._initialStatus();
    this._outbox = [];
    this._nextSeq = 0;
    this._cancelRequested = false;
    this._listeners = new Set();
    this._persistChain = Promise.resolve();

    // Restored asynchronously (chrome.storage is always async, even for a
    // service worker that was just woken up). Callers that care -- index.js,
    // and any test exercising restore-from-storage -- `await runner.ready`.
    this.ready = this._restore();
  }

  static _initialStatus() {
    return {
      state: "idle",
      message: "",
      pages_completed: 0,
      jobs_saved: 0,
      cards_seen: 0,
      duplicates: 0,
      failed: 0,
      failed_jobs: [],
      duplicate_jobs: [],
      anomalies: 0,
      retries: 0,
      retry_events: [],
      current: null,
      total_pages: null,
      estimated_jobs: null,
      events: [],
      error: null,
      reason: null,
      started_at: null,
      finished_at: null,
    };
  }

  get isRunning() {
    return RUNNING_STATES.has(this._status.state);
  }

  /** JSON-serialisable snapshot, including the current version and the
   * outbox (so index.js can push both to a freshly (re)connected bridge
   * port in one message -- see the protocol notes in index.js). */
  status() {
    return { ...this._status, events: [...this._status.events], version: this._version, outbox: [...this._outbox] };
  }

  /** Replaces wait_for_change()/SSE: `listener(snapshot)` fires on every
   * bump. Returns an unsubscribe function. */
  subscribe(listener) {
    this._listeners.add(listener);
    return () => this._listeners.delete(listener);
  }

  /** Test/shutdown hook: resolves once the latest persist() has landed. */
  async flush() {
    await this._persistChain;
  }

  // -- persistence --------------------------------------------------------

  async _restore() {
    if (!this.storage) return;
    let data;
    try {
      data = await this.storage.get(this._storageKey);
    } catch {
      return; // a storage read failing must never crash construction
    }
    const saved = data && data[this._storageKey];
    if (!saved) return;
    const { events = [], version = 0, ...rest } = saved.status || {};
    this._status = { ...ScrapeRunner._initialStatus(), ...rest, events };
    this._version = version;
    this._outbox = Array.isArray(saved.outbox) ? saved.outbox : [];
    this._nextSeq = saved.nextSeq || 0;
  }

  _persist() {
    if (!this.storage) return;
    const payload = { status: this.status(), outbox: this._outbox, nextSeq: this._nextSeq };
    delete payload.status.outbox; // stored separately, avoid duplicating it
    // Chained (not fire-and-forget in parallel) so writes land in order even
    // if the caller doesn't await; a failing write is logged, never thrown.
    this._persistChain = this._persistChain
      .then(() => this.storage.set({ [this._storageKey]: payload }))
      .catch((exc) => console.warn("[CoopJobs] could not persist scraper status:", exc));
  }

  _bump() {
    this._version += 1;
    this._persist();
    const snapshot = this.status();
    for (const listener of this._listeners) {
      try {
        listener(snapshot);
      } catch (exc) {
        console.warn("[CoopJobs] scraper status subscriber threw:", exc);
      }
    }
  }

  _appendEvent(message, level = "info") {
    this._status.events.push({ time: now(), level, message });
    if (this._status.events.length > EVENT_LOG_MAX) {
      this._status.events.splice(0, this._status.events.length - EVENT_LOG_MAX);
    }
  }

  /** Like runner.py's _update: patch fields, bump, no log line. */
  update(fields) {
    Object.assign(this._status, fields);
    this._bump();
  }

  /** Port of runner.py's _set_state. */
  _setState(state, message, { level = "info", ...fields } = {}) {
    if (RUNNING_STATES.has(state) && this._status.state === "cancelling") {
      // An abort is already under way; a running state arriving late must
      // not undo it. Terminal states always apply -- that is how a run ends.
      return;
    }
    Object.assign(this._status, { state, message, ...fields });
    if (!RUNNING_STATES.has(state)) this._status.current = null;
    this._appendEvent(message, level);
    this._bump();
  }

  // -- lifecycle ------------------------------------------------------------

  /** Refused when a run is active, or when unacknowledged jobs from a prior
   * run are still in the outbox (draining those takes priority). The
   * database-not-empty check lives on the web side (see extensionBridge.js /
   * api.js), which is the only side with access to the SQLite database. */
  start() {
    if (this.isRunning) {
      throw new AlreadyRunning("A scrape is already running.");
    }
    if (this._outbox.length > 0) {
      throw new OutboxNotEmpty("Jobs from the previous run are still being delivered.");
    }
    this._cancelRequested = false;
    this._status = ScrapeRunner._initialStatus();
    this._status.state = "starting";
    this._status.message = "Starting the scraper browser.";
    this._status.started_at = now();
    this._appendEvent(this._status.message);
    this._bump();
    return this.status();
  }

  /** Called once the service worker has opened the portal window/tab and a
   * port to its content script exists -- the equivalent of the Python worker
   * thread reaching its first _set_state("waiting_for_login", ...) call. */
  markWaitingForLogin() {
    this._setState("waiting_for_login", "Log in to the portal in the browser window that opened.");
  }

  /** Port of runner.py's cancel(). Quitting the actual browser/tab is
   * index.js's job (it owns the chrome.tabs/windows APIs); this only flips
   * the state so every reader -- including a stale response racing the
   * worker -- already sees "cancelling". */
  cancel() {
    this._cancelRequested = true;
    if (RUNNING_STATES.has(this._status.state)) {
      this._setState("cancelling", "Aborting the scrape.", { level: "warn" });
    }
    return this.status();
  }

  /** Wipes a finished run's report (last_run/finished_at/counters/events),
   * called when the web app deletes all jobs so a stale "Scrape complete"
   * report doesn't survive an empty database. Refused while a scrape is
   * actually in flight -- there is nothing to reset yet, and clearing state
   * out from under a running scraper would corrupt its accounting. */
  reset() {
    if (this.isRunning) return this.status();
    this._status = ScrapeRunner._initialStatus();
    // The outbox and _nextSeq are persisted and restored together with
    // _status (see _persist/_restore), and status() rides the outbox along
    // in every snapshot the web app drains on connect. Clearing _status
    // alone would leave unacknowledged jobs from the wiped-out run sitting
    // in the outbox, ready to be re-inserted into the database the user
    // just emptied -- so both must be cleared in the same breath.
    this._outbox = [];
    // Resetting _nextSeq to 0 is only safe because the outbox is emptied
    // right above: with no old entries left, no future seq can collide
    // with one still in flight. These two lines look independent but
    // are not -- don't reset one without the other.
    this._nextSeq = 0;
    this._bump();
    return this.status();
  }

  // -- reporting from the portal content script ----------------------------

  /** Port of runner.py's on_ready. */
  handleReady() {
    this._setState("scraping", "Portal login detected. Scraping jobs now.");
  }

  _applyAccounting(accounting = {}) {
    if (!accounting) return;
    if ("cardsSeen" in accounting) this._status.cards_seen = accounting.cardsSeen;
    if ("duplicates" in accounting) this._status.duplicates = accounting.duplicates;
    if ("failed" in accounting) this._status.failed = accounting.failed;
    if ("failedJobs" in accounting) this._status.failed_jobs = accounting.failedJobs.slice(0, 50);
    if ("duplicateJobs" in accounting) this._status.duplicate_jobs = accounting.duplicateJobs.slice(0, 50);
  }

  /** Port of runner.py's on_event. `accounting` is the scraper's own
   * cardsSeen/duplicates/failed/failedJobs counters as read by the content
   * script in the same tick it built `event` -- the equivalent of Python's
   * sync_accounting(), which could read scraper.* directly because runner.py
   * and JobPortalScraper shared a thread; here they don't, so the content
   * script (the only side that still holds the live scraper instance) sends
   * a fresh snapshot with every event instead. */
  handleEvent(event, accounting = {}) {
    const kind = event.type || "";
    this._applyAccounting(accounting);

    if (kind === "pages_found") {
      const totalPages = event.total_pages ?? null;
      const cardsPerPage = event.cards_per_page || 0;
      this._status.total_pages = totalPages;
      this._status.estimated_jobs = totalPages && cardsPerPage ? totalPages * cardsPerPage : null;
    } else if (kind === "card") {
      this._status.current = {
        page: event.page,
        index: event.index,
        total: event.total,
        title: event.title || "",
      };
    } else if (kind === "scraped" && event.job) {
      this._enqueueOutboxEntry(event);
      this._bump(); // accounting/outbox changed even though nothing is logged yet
      return;
    } else if (kind === "retry") {
      this._status.retries += 1;
      const msg = eventMessage(event);
      this._status.retry_events.push(msg);
      if (this._status.retry_events.length > 50) {
        this._status.retry_events.splice(0, this._status.retry_events.length - 50);
      }
    }

    if (kind !== "card") {
      const level = EVENT_LEVELS[kind] || "info";
      this._appendEvent(eventMessage(event), level);
    }
    this._bump();
  }

  /** Port of runner.py's on_page: counters and the status message only --
   * jobs were already written by their own `scraped` events (main sweep and
   * retry pass alike), so this never touches jobs_saved or the outbox. */
  handlePage(pageNumber, jobsCount, accounting = {}) {
    this._applyAccounting(accounting);
    this._status.pages_completed += 1;
    this._status.message = `Finished page ${pageNumber} (${jobsCount} job(s)).`;
    this._bump();
  }

  /** Port of the tail of runner.py's _run try block: the scraper's run()
   * resolved without throwing. `failedJobs` must be the FINAL, post-retry
   * list (JobPortalScraper.failed after run() returns) -- see the module
   * docstring in portal.js and runner.py alike. */
  handleFinished({ cardsSeen = 0, duplicates = 0, failedJobs = [], duplicateJobs = [] } = {}) {
    const cappedFailedJobs = failedJobs.slice(0, 50);
    const cappedDuplicateJobs = duplicateJobs.slice(0, 50);
    const jobsSaved = this._status.jobs_saved;
    this._setState(
      "completed",
      `Saved ${jobsSaved} of ${cardsSeen} jobs (${duplicates} duplicates, ${failedJobs.length} failed).`,
      {
        cards_seen: cardsSeen,
        duplicates,
        failed: failedJobs.length,
        failed_jobs: cappedFailedJobs,
        duplicate_jobs: cappedDuplicateJobs,
        anomalies: failedJobs.length,
        finished_at: now(),
      },
    );
  }

  /** Port of runner.py's except Cancelled / except (BrowserClosed, ...) /
   * except Exception cascade. `name` is the JS error class name the content
   * script's `finished`-with-error message carries (Cancelled, TabClosed, or
   * anything else); `message` is its .message. */
  // `reason` is a machine-readable companion to `message`, so the UI (and
  // tests) don't have to string-match "Browser window was closed." to tell
  // a real cancel apart from a closed tab.
  handleError({ name = "", message = "" } = {}) {
    if (name === "Cancelled" || this._cancelRequested) {
      this._setState("cancelled", "Scrape was cancelled.", {
        level: "warn",
        reason: "user_cancelled",
        finished_at: now(),
      });
      return;
    }
    if (name === "TabClosed" || isBrowserClosedMessage(message)) {
      this._setState("cancelled", "Browser window was closed.", {
        level: "warn",
        reason: "browser_closed",
        finished_at: now(),
      });
      return;
    }
    this._setState("failed", message, { level: "error", error: message, reason: "failed", finished_at: now() });
  }

  /** Called by index.js when chrome.tabs.onRemoved (or a port disconnect)
   * fires for the tracked portal tab while a run is active -- there is no
   * content-script message at all in that case, since the tab is gone. */
  handleTabClosed() {
    if (!RUNNING_STATES.has(this._status.state)) return;
    this.handleError({ name: "TabClosed", message: "Browser window was closed." });
  }

  // -- outbox: jobs in flight to the web app's database --------------------

  _enqueueOutboxEntry(event) {
    const seq = this._nextSeq;
    this._nextSeq += 1;
    const entry = {
      seq,
      job: event.job,
      page: event.page,
      index: event.index,
      total: event.total,
      title: event.title || "",
    };
    this._outbox.push(entry);
    this._sendToTransport(entry);
  }

  _sendToTransport(entry) {
    if (!this.transport) return;
    try {
      this.transport.sendJob(entry);
    } catch (exc) {
      console.warn("[CoopJobs] could not hand a scraped job to the transport:", exc);
    }
  }

  /** Re-sends every still-unacknowledged outbox entry, in order -- called by
   * index.js once a bridge port (re)connects, draining what built up while
   * the web app tab was closed or unreachable. */
  resendOutbox() {
    for (const entry of this._outbox) this._sendToTransport(entry);
  }

  outbox() {
    return [...this._outbox];
  }

  /** The web app's ack for one outbox entry: `{inserted, error}`, the same
   * accounting runner.py's save_job produced synchronously. Unknown/already
   * -acked seqs are a silent no-op, since the web side also dedupes by seq
   * and may re-ack after a reconnect. */
  ack(seq, { inserted = 0, error = "" } = {}) {
    const index = this._outbox.findIndex((entry) => entry.seq === seq);
    if (index === -1) return this.status();
    const [entry] = this._outbox.splice(index, 1);

    let saveError = "";
    if (error) saveError = error;
    else if (!inserted) saveError = "the database already has that job number";

    if (!saveError) this._status.jobs_saved += 1;
    const level = saveError ? "error" : "info";
    this._appendEvent(
      eventMessage(
        { type: "scraped", page: entry.page, index: entry.index, total: entry.total, title: entry.title },
        saveError,
      ),
      level,
    );
    this._bump();
    return this.status();
  }
}
